pkg/operator/ceph/csi/spec.go

```diff
@@ -662,3 +688,265 @@ func (r *ReconcileCSI) validateCSIVersion(ownerInfo *k8sutil.OwnerInfo) (*CephCS
 	}
 	return version, nil
 }
+
+type migrationJobReport struct {
+	nodeName string
+	err      error
+}
+
+func (r *ReconcileCSI) setupCSINetwork() error {
+	// Setup of the CSI Network depends on if multus is being used.
+	var isMultusApplied bool
+	var publicNetwork string
+	var ok bool
+
+	cephClusters := &cephv1.CephClusterList{}
+	err := r.client.List(r.opManagerContext, cephClusters, &client.ListOptions{})
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			logger.Debug("no ceph cluster found not deploying ceph csi driver")
+			return nil
+		}
+		// Error reading the object - requeue the request.
+		return errors.Wrap(err, "failed to list ceph clusters")
+	}
+
+	for _, cephCluster := range cephClusters.Items {
+		// Assuming one CephCluster in the r.opConfig.OperatorNamespace
+		if cephCluster.Spec.Network.IsMultus() {
+			if publicNetwork, ok = cephCluster.Spec.Network.Selectors["public"]; !ok {
+				logger.Debug("public network not provided; not performing multus configuration.")
+				return nil
+			}
+			isMultusApplied = true
+			break
+		}
+	}
+
+	if !isMultusApplied {
+		logger.Debug("multus not used; no additional network configuration necessary.")
+		return nil
+	}
+
+	// Populate the host network namespace with a multus-connected interface.
+	tp := templateParam{
+		Namespace: r.opConfig.OperatorNamespace,
+	}
+
+	tp.MultusNetwork = publicNetwork
+	tp.MultusImage = r.opConfig.Image
+
+	multusHostSetup, err := templateToDaemonSet(csiMultusSetup, MultusDaemonsetTemplatePath, tp)
+	if err != nil {
+		return errors.Wrap(err, "failed to get daemonset template to set up the csi network")
+	}
+
+	// Applying affinity and toleration to multus daemonset.
+	// The multus network must be configured on every node that csi plugin pods run on.
+	pluginTolerations := getToleration(r.opConfig.Parameters, pluginTolerationsEnv, []corev1.Toleration{})
+	pluginNodeAffinity := getNodeAffinity(r.opConfig.Parameters, pluginNodeAffinityEnv, &corev1.NodeAffinity{})
+	applyToPodSpec(&multusHostSetup.Spec.Template.Spec, pluginNodeAffinity, pluginTolerations)
+
+	err = k8sutil.CreateDaemonSet(csiMultusSetup, r.opConfig.OperatorNamespace, r.context.Clientset, multusHostSetup)
+	if err != nil {
+		return errors.Wrap(err, "failed to start daemonset to set up the csi network")
+	}
+
+	// Wait for daemonset pods to come up.
+	// Will poll daemonset to ensure pods are available for interface migration.
+	err = wait.Poll(5*time.Second, 30*time.Second, func() (bool, error) {
```

Comment:
So if the pods don't start on all nodes within 30s the csi driver start fails completely? Waiting for pods can be a long running operation, especially if images are taking time to be pulled. Can we add this to a goroutine and allow it to complete asynchronously? Or it is really a blocking issue that we need to wait for the daemonset pods? 

Or for example, what if one of the nodes is unavailable? Perhaps it has been marked unschedulable or drained for a restart. Will the daemonset status allow completion of the other pods? Or will this be blocked until the node comes back up?

What happens if a new node is added to the cluster? Do we watch for that event and configure multus for the new pod as well?